engine/src/main/java/org/hibernate/validator/internal/engine/valueextraction/ValueExtractorManager.java

```diff
@@ -347,6 +351,124 @@ private static boolean isClassPresent(String className, boolean fallbackOnTCCL)
 		}
 	}
 
+	private final class ValueExtractorCache {
+
+		private final ConcurrentHashMap<ValueExtractorCacheKey, Set<ValueExtractorDescriptor>> valueExtractorCache;
+		private final Set<ValueExtractorCacheKey> nonContainerTypes;
+
+		public ValueExtractorCache() {
+			this.valueExtractorCache = new ConcurrentHashMap<>();
+			this.nonContainerTypes = new HashSet<>();
+		}
+
+		public Set<ValueExtractorDescriptor> getValueExtractors( Set<ValueExtractorDescriptor> valueExtractorCandidates,
```

Comment:
So in my mind, I wouldn't have the candidates here. I would simply have the runtime type and the type parameter.

The whole notion of candidates would be removed. I thought that you could do it in 2 steps but maybe you can't.

Reply:
yes, I just used the candidates here to try and see how the map would behave in general. As this value will be cached there shouldn't be a difference if we look just through the candidates or through all extractors, as in my understanding it should slow down only the first call and then the cache will be used, right ?

Reply:
I also thought at first to store just a resulting value extractor 
```java
ConcurrentHashMap<ValueExtractorCacheKey, ValueExtractorDescriptor> valueExtractorCache
```
but then it wouldn't allow us to know if there are two or more extractors and we should throw an error. The alternative might be to have
```java
private final ConcurrentHashMap<ValueExtractorCacheKey, Supplier<ValueExtractorDescriptor>> valueExtractorCache;
```
and then if there are more than 1 extractor we could store something like
```java
() -> throw LOG.getUnableToGetMostSpecificValueExtractorDueToSeveralMaximallySpecificValueExtractorsDeclaredException( valueType,
					ValueExtractorHelper.toValueExtractorClasses( maximallySpecificContainerElementCompliantValueExtractors )
			);
```
but I wasn't sure that that would be a good idea...

Reply:
Yes, storing a Set looks like the right thing to do.